pkg/controller/infinispan/infinispan_controller.go

```diff
@@ -235,6 +233,22 @@ func (r *ReconcileInfinispan) Reconcile(request reconcile.Request) (reconcile.Re
 		return *result, err
 	}
 
+	var keystoreSecret *corev1.Secret
+	if infinispan.IsEncryptionEnabled() {
+		keystoreSecret = &corev1.Secret{}
+		if result, err := kube.LookupResource(infinispan.GetKeystoreSecretName(), infinispan.Namespace, keystoreSecret, r.client, reqLogger); result != nil {
```

Comment:
I've a doubt here: what happens if I delete an encryption secret for a running Infinispan? Infinispan continue to work but the reconciliation logic is not executed until the secret will be available again?

Reply:
Yes, but I'm not sure there is anyway we can avoid this? The secret is a dependency of the Infinispan cluster.

Reply:
If secrets are needed only by the StatefulSet, how about do the lookup only when the StatefulSet check is not found?

Reply:
The reason we now need to retrieve the secrets beforehand is to support reconcilliation of the key/truststore secrets. So if a user changes the values of either secret, then the cluster is redeployed with the new certs. For this to work, we need to retrieve the secrets before `reconcileContainerConf` as well as when creating a new StatefulSet.

Reply:
yep, but if secrets are not found there's no need to check for changes, unless user explicitly disabled keystore/trustore usage and operator can work until secrets reappear or statefulset needs to be recreated due to other changes.
anyway I didn't check the impact of this on the code and maybe I'm just describing an marginal corner case

Reply:
If the secrets aren't found, then the creation of the StatefulSet will be blocked anyway as the secrets are mounted as volumes.

Reply:
that's fine, but the case I'm trying to describe is this: the cluster is up and running the secrets disappear (delete or lost) and no other changes to the CR are applied (i.e. client auth is still on)
With the current implementation the reconcile loop seems to stop here. I'm wondering if it can go on with the rest of the reconciliation policy (cluster status update, cross site coordinator...) while the secrets are missing.
I mean to restart the cluster (statefulset redeploy) a set of new secrets is needed, but in the while the cluster is working with the old one and it can be managed by the operator.

Reply:
@rigazilla probably reconcile not finished in case of Secret delete/update as we are listening on the Secrets as secondary watch objects. Probably I missed something, but this case need to be validated. That's the reason why we are setting owner reference to the custom defined secrets.

Reply:
> With the current implementation the reconcile loop seems to stop here. I'm wondering if it can go on with the rest of the reconciliation policy (cluster status update, cross site coordinator...) while the secrets are missing.
> I mean to restart the cluster (statefulset redeploy) a set of new secrets is needed, but in the while the cluster is working with the old one and it can be managed by the operator.

I'm not sure there's an advantage in this though, as if the secret doesn't exist then the cluster is effictively dead as it's outdated. Allowing the cluster to continue operating might be more confusing for the user, as they won't realise there is an issue unless they check the operator logs. If any of the required secrets are not found on reconcilliation, we could delete the old StatefulSet to prevent clients continuing to execute against the cluster?

It could also be worth adding a condition to signal whether all secrets are available or not?

Reply:
> I'm not sure there's an advantage in this though, as if the secret doesn't exist then the cluster is effictively dead as it's outdated. Allowing the cluster to continue operating might be more confusing for the user, as they won't realise there is an issue unless they check the operator logs. If any of the required secrets are not found on reconcilliation, we could delete the old StatefulSet to prevent clients continuing to execute against the cluster?
> 

This could be an option: cluster is actually working with an obsolete configuration. I'm not 100% happy on killing a working cluster, but yeah, maybe it's better to keep things clean.
This make impossible for the user to recover from an unwanted delete: i.e. redeploy the same secret. I'm not sure if this is something doable anyway.

Reply:
> This could be an option: cluster is actually working with an obsolete configuration. I'm not 100% happy on killing a working cluster, but yeah, maybe it's better to keep things clean.

Maybe we should keep things as is for now? We can then put more thought into this for the next release.

> This make impossible for the user to recover from an unwanted delete: i.e. redeploy the same secret. I'm not sure if this is something doable anyway.

AFAIK in the current code this wouldn't actually result in the new secret being used until another update, i.e. spec change, would cause the StatefulSet to be recreated.

Reply:
> Maybe we should keep things as is for now? We can then put more thought into this for the next release.
> 

fine, overall looks good to me. Let's go with this and see if users provide feedback for more fancy stuff :)